src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -26,26 +26,75 @@
+ * (based on a subset of the built-in types provided by the C language), found in this interface;
+ * a failure to do so might result in linkage errors, given that linking requires additional classification
+ * information to determine, for instance, how arguments should be loaded into registers during a
+ * foreign function call.</p>
+ *
+ * <p>Implementations of this interface support the following primitive carrier types:
+ * {@code byte}, {@code short}, {@code char}, {@code int}, {@code long}, {@code float},
+ * and {@code double}, as well as {@link MemoryAddress} for passing pointers, and
+ * {@link MemorySegment} for passing structs and unions. Finally, the {@link VaList}
+ * carrier type can be used to match the native {@code va_list} type.</p>
+ *
+ * <p>The function descriptor used in linking contains a memory layout to match each carrier type.
+ * There are some restrictions on the carrier type and memory layout combinations that are allowed:</p>
+ * <ul>
+ *   <li>If a primitve type is used as a carrier type, the corresponding
+ *   memory layout must be a {@code ValueLayout}, and the bit size of the layout must match that of the carrier type
+ *   (see {@link Integer#SIZE} and similar fields in other primitive wrapper classes).</li>
+ *
+ *   <li>If the carrier type is {@code MemoryAddress}, then the corresponding memory layout must be a
+ *   {@code ValueLayout}, and its bit size must match the platform's address size (see {@link MemoryLayouts#ADDRESS}).
+ *   For this purpose, {@link CLinker#C_POINTER} can  be used</li>
+ *
+ *   <li>If the carrier type is {@code MemorySegment}, then the corresponding memory layout must be a
+ *   {@code GroupLayout}</li>
+ *
+ *   <li>If the carrier type is {@code VaList}, then the corresponding memory layout must be
+ *   {@link CLinker#C_VA_LIST}</li>
+ * </ul>
+ *
+ * <p>Variadic functions, declared in C either with a trailing ellipses ({@code ...}) at the end of the formal parameter
+ * list or with an empty formal parameter list, are not supported directly. It is not possible to create a method handle
+ * that takes a variable number of arguments, and neither is it possible to create an upcall stub wrapping a method
+ * handle that accepts a variable number of arguments. However, for downcalls only, it is possible to link a native
+ * variadic function by using a <em>specialized</em> method type and function descriptor: for each argument that is to be
+ * passed as a variadic argument, an explicit carrier type and memory layout must be present in the method type and
+ * function descriptor when linking the function. Furthermore, as memory layouts corresponding to variadic arguments in
+ * a function descriptor must contain additional classification information, it is required that
+ * {@link #asVarArg(MemoryLayout)} is used to create the memory layouts for each parameter corresponding to a variadic
+ * argument in a specialized function descriptor</p>
  */
```

Comment:
There should be the usual comment about sealedness and immutability - see MemoryLayout